src/chain_actions.cc

```diff
@@ -336,6 +337,149 @@ namespace dd
     model_out.add("predictions", cvad);
   }
 
+  cv::Scalar bbox_palette[]
+      = { { 82, 188, 227 }, { 196, 110, 49 }, { 39, 54, 227 },
+          { 68, 227, 81 },  { 77, 157, 255 }, { 255, 112, 207 },
+          { 240, 228, 65 }, { 94, 242, 151 }, { 236, 121, 242 },
+          { 28, 77, 120 } };
+  size_t bbox_palette_size = 10;
+
+  void ImgsDrawBBoxAction::apply(APIData &model_out, ChainData &cdata)
+  {
+    std::vector<APIData> vad = model_out.getv("predictions");
+    APIData input_ad = model_out.getobj("input");
+
+    std::vector<cv::Mat> imgs
+        = model_out.getobj("input").get("imgs").get<std::vector<cv::Mat>>();
+    std::vector<std::pair<int, int>> imgs_size
+        = model_out.getobj("input")
+              .get("imgs_size")
+              .get<std::vector<std::pair<int, int>>>();
+    std::vector<cv::Mat> rimgs;
+    std::vector<std::string> uris;
+    auto pred_body = DTO::PredictBody::createShared();
+
+    bool save_img = _params->save_img;
+    int ref_thickness = _params->thickness;
+
+    std::string save_path = _params->save_path->std_str();
+    if (!save_path.empty())
+      save_path += "/";
+
+    for (size_t i = 0; i < vad.size(); i++)
+      {
+        std::string uri = vad.at(i).get("uri").get<std::string>();
+        uris.push_back(uri);
+
+        int im_cols = imgs.at(i).cols;
+        int im_rows = imgs.at(i).rows;
+        int orig_cols = imgs_size.at(i).second;
+        int orig_rows = imgs_size.at(i).first;
+
+        std::vector<APIData> ad_cls = vad.at(i).getv("classes");
+        cv::Mat rimg = imgs.at(i).clone();
```

Comment:
Do we need/want to support GPUMat as well ?

Reply:
Since we cannot yet encode video on the GPU there is no use for a GPUMat at this stage. So the support for GPUMat can be left for another PR.

Reply:
OK understood